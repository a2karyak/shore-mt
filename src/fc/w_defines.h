/*<std-header orig-src='shore' incl-file-exclusion='W_DEFINES_H' no-defines='true'>

 $Id: w_defines.h,v 1.3 2008/05/31 05:03:26 nhall Exp $

SHORE -- Scalable Heterogeneous Object REpository

Copyright (c) 1994-99 Computer Sciences Department, University of
                      Wisconsin -- Madison
All Rights Reserved.

Permission to use, copy, modify and distribute this software and its
documentation is hereby granted, provided that both the copyright
notice and this permission notice appear in all copies of the
software, derivative works or modified versions, and any portions
thereof, and that both notices appear in supporting documentation.

THE AUTHORS AND THE COMPUTER SCIENCES DEPARTMENT OF THE UNIVERSITY
OF WISCONSIN - MADISON ALLOW FREE USE OF THIS SOFTWARE IN ITS
"AS IS" CONDITION, AND THEY DISCLAIM ANY LIABILITY OF ANY KIND
FOR ANY DAMAGES WHATSOEVER RESULTING FROM THE USE OF THIS SOFTWARE.

This software was developed with support by the Advanced Research
Project Agency, ARPA order number 018 (formerly 8230), monitored by
the U.S. Army Research Laboratory under contract DAAB07-91-C-Q518.
Further funding for this work was provided by DARPA through
Rome Research Laboratory Contract No. F30602-97-2-0247.

*/

#ifndef W_DEFINES_H
#define W_DEFINES_H

/*  -- do not edit anything above this line --   </std-header>*/

#include "shore.def"
/* shore-config.h does not have duplicate-include protection, but
   this file does, and we don't include shore-config.h anywhere else*/
#include "shore-config.h"

// OK, shore.def might or might not define SM_PAGESIZE
// If not, then shore-config.h defines it to 8K
// shore-config.h can define LARGE_PAGE based on config option --enable-largepages
// Now we need to force largepages if SM_PAGESIZE is too big
#if SM_PAGESIZE > 32768
#define LARGE_PAGE
#endif

/* Adapt shore to namespaces */
#if defined(__cplusplus) && !defined(USE_CPP_NAMESPACE_NO)
#ifdef _MSC_VER
namespace std {};
#endif
using namespace std;
#endif

/*<std-footer incl-file-exclusion='W_DEFINES_H'>  -- do not edit anything below this line -- */

#endif          /*</std-footer>*/
