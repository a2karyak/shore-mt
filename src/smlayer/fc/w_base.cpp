/*<std-header orig-src='shore'>

 $Id: w_base.cpp,v 1.45 2000/02/01 23:20:49 bolo Exp $

SHORE -- Scalable Heterogeneous Object REpository

Copyright (c) 1994-99 Computer Sciences Department, University of
                      Wisconsin -- Madison
All Rights Reserved.

Permission to use, copy, modify and distribute this software and its
documentation is hereby granted, provided that both the copyright
notice and this permission notice appear in all copies of the
software, derivative works or modified versions, and any portions
thereof, and that both notices appear in supporting documentation.

THE AUTHORS AND THE COMPUTER SCIENCES DEPARTMENT OF THE UNIVERSITY
OF WISCONSIN - MADISON ALLOW FREE USE OF THIS SOFTWARE IN ITS
"AS IS" CONDITION, AND THEY DISCLAIM ANY LIABILITY OF ANY KIND
FOR ANY DAMAGES WHATSOEVER RESULTING FROM THE USE OF THIS SOFTWARE.

This software was developed with support by the Advanced Research
Project Agency, ARPA order number 018 (formerly 8230), monitored by
the U.S. Army Research Laboratory under contract DAAB07-91-C-Q518.
Further funding for this work was provided by DARPA through
Rome Research Laboratory Contract No. F30602-97-2-0247.

*/

#include "w_defines.h"

/*  -- do not edit anything above this line --   </std-header>*/

#define W_SOURCE

#ifdef __GNUG__
#pragma implementation "w_base.h"
#endif

#ifdef _MSC_VER
/* Undefined for our purecall definition */
#define	_purecall	libc_purecall
#endif

#include <w_base.h>
#include <unix_error.h>


/*--------------------------------------------------------------*
 *  constants for w_base_t                                      *
 *--------------------------------------------------------------*/
const w_base_t::int1_t    w_base_t::int1_max  = 0x7f;
const w_base_t::int1_t    w_base_t::int1_min  = (const w_base_t::int1_t) 0x80u;
const w_base_t::uint1_t   w_base_t::uint1_max = 0xff;
const w_base_t::uint1_t   w_base_t::uint1_min = 0x0;
const w_base_t::int2_t    w_base_t::int2_max  = 0x7fff;
const w_base_t::int2_t    w_base_t::int2_min  = (const w_base_t::int2_t) 0x8000u;
const w_base_t::uint2_t   w_base_t::uint2_max = 0xffff;
const w_base_t::uint2_t   w_base_t::uint2_min = 0x0;
const w_base_t::int4_t    w_base_t::int4_max  = 0x7fffffff;
const w_base_t::int4_t    w_base_t::int4_min  = 0x80000000;
const w_base_t::uint4_t   w_base_t::uint4_max = 0xffffffff;
const w_base_t::uint4_t   w_base_t::uint4_min = 0x0;

#ifdef _MSC_VER
#	define LONGLONGCONSTANT(i) i
#	define ULONGLONGCONSTANT(i) i
#else
#	define LONGLONGCONSTANT(i) i##LL
#	define ULONGLONGCONSTANT(i) i##ULL
#endif /* _MSC_VER */

const w_base_t::uint8_t   w_base_t::uint8_max =
				ULONGLONGCONSTANT(0xffffffffffffffff);
const w_base_t::uint8_t   w_base_t::uint8_min =
				ULONGLONGCONSTANT(0x0);
const w_base_t::int8_t   w_base_t::int8_max =
				LONGLONGCONSTANT(0x7fffffffffffffff);
const w_base_t::int8_t   w_base_t::int8_min =
				LONGLONGCONSTANT(0x8000000000000000);

ostream&
operator<<(ostream& o, const w_base_t&)
{
    w_base_t::assert_failed("w_base::operator<<() called", __FILE__, __LINE__);
    return o;
}

void
w_base_t::assert_failed(
    const char*		desc,
    const char*		file,
    uint4_t		line)
{
#ifdef notyet
	/* generate a normal RC that makes assertion failure look just 
	   like any other fatal error.  On-stack hack is used
	   to prevent fastnew code from being used. */
	w_error_t	er(file, line, fcASSERT, 0);
	w_rc_t		e(&er);
	W_COERCE(e);
	e.reset();
#else
	/* make the error look something like an RC in the meantime. */
	cerr << "assertion failure: " << desc << endl
		<< "1. error in "
		<< file << ':' << line
		<< " Assertion failed" << endl
		<< "\tcalled from:" << endl
		<< "\t0) " << file << ':' << line
		<< endl << flush;
	abort();
#endif
}


/* XXX compilers and/or environments which need these */
#if defined(_MSC_VER) 

/* XXX These output operators only follow base and showbase requests
   on win32 and visual c++.  Which is better than before, so 
   I'm working on it. */

ostream &operator<<(ostream &o, const w_base_t::uint8_t &t)
{
	char buf[24];

#ifdef _MSC_VER
	char	*bufp = buf;
	int	base = 10;	/* the default */
	if (o.flags() & ios::hex) {
		base = 16;
		if (o.flags() & ios::showbase) {
			*bufp++ = '0';
			*bufp++ = 'x';
		}
	}
	else if (o.flags() & ios::oct) {
		base = 8;
		if (t && o.flags() & ios::showbase)
			*bufp++ = '0';
	}

	_ui64toa(t, bufp, base);

	bufp = buf;
#else
	char *bufp = &buf[sizeof(buf)-1];
	*bufp = '\0';
	bufp =  ulltostr(t, bufp);
#endif

	o << bufp;
	return o;
}

ostream &operator<<(ostream &o, const w_base_t::int8_t &t)
{
	char buf[24];

#ifdef _MSC_VER
	char	*bufp = buf;
	int	base = 10;	/* the default */
	if (o.flags() & ios::hex) {
		base = 16;
		if (o.flags() & ios::showbase) {
			*bufp++ = '0';
			*bufp++ = 'x';
		}
	}
	else if (o.flags() & ios::oct) {
		base = 8;
		if (t && o.flags() & ios::showbase)
			*bufp++ = '0';
	}
	
	
	/* XXX workaround for formatting error with i64toa, the
	   fix will work correctly if i64toa is fixed. */
	if (base == 10 && t < 0)
		*bufp++ = '-';

	_ui64toa((t < 0 && base == 10) ? -t : t, bufp, base);

	bufp = buf;
#else
	/* XXX workaround for formatting error with lltostr, the
	   fix will work correctly if lltostr is fixed. */
	char *bufp = &buf[sizeof(buf)-1];
	*bufp = '\0';
	bufp =  ulltostr(t < 0 ? -t : t, bufp);
	if (t < 0 && bufp)
		*--bufp = '-';
#endif
	o << bufp;
	return o;
}

istream &
operator>>(istream& i, w_base_t::uint8_t &t)
{
	bool	rangerr=false;
	return w_base_t::_scan_uint8(i, t, true, false, rangerr);
}

istream &
operator>>(istream& i, w_base_t::int8_t &t)
{
	bool	rangerr=false;
	w_base_t::uint8_t u;
	istream& res = w_base_t::_scan_uint8(i, u, true, true, rangerr);
	t = w_base_t::int8_t(u);
	return res;
}
#endif

#ifdef _MSC_VER

#include <string.h>
#include <ios.h>
#include <strstrea.h>
typedef long fmtflags;

#else

#include <strstream.h>
typedef ios::fmtflags  fmtflags;

#endif


/* USE_OUR_IMPLEMENTATION lets us test the
 * code in w_input.cpp
 * on Solaris or Linux or NetBSD 
 */
#undef USE_OUR_IMPLEMENTATION

#ifdef _MSC_VER
#define USE_OUR_IMPLEMENTATION
#endif


/* name is local to this file */
static w_base_t::uint8_t	
__strtou8(		
    const char *str,
    char **endptr,
    int 	base,
    bool  is_signed
)
{

#if defined(_MSC_VER) || defined(USE_OUR_IMPLEMENTATION)

    istrstream 		tmp(VCPP_BUG_1 str, strlen(str));
    streampos 		b =  tmp.tellg();

    fmtflags f = tmp.flags();
    f = f & ~(ios::basefield);
    switch(base) {
       case 0:
	    break;
       case 10:
	    f |= ios::dec;
	    break;
       case 8:
	    f |= ios::oct;
	    break;
       case 16:
	    f |= ios::hex;
	    break;
       default:
	    /*
	    cerr << "w_base_t::__strtou8 : base "
		    << base << " not supported" << endl;
	    W_FATAL(fcNOTIMPLEMENTED);
	    */
	    errno = EINVAL;
	    return 0;
	    break;
    }
    // strto*() permit leading white space
    f |= ios::skipws;
    tmp.flags(f);

#ifdef W_DEBUG
    if(tmp.flags() != f) {
	w_assert3(0);
    }
#endif

    w_base_t::uint8_t	u8;
    bool			rangerr=false;
    w_base_t::_scan_uint8(tmp, u8, false, is_signed, rangerr);

    if(rangerr) {
	errno = ERANGE;
    } 
    if(endptr) {
	streampos 		e =  tmp.tellg();
	*endptr = (char *)&str[e - b];
    }
    return u8;

#elif defined(__NetBSD__)
    return is_signed? ::strtoq(str, endptr, base): ::strtouq(str, endptr, base);
#elif defined(Alpha)
    return is_signed? strtol(str, endptr, base): strtoul(str, endptr, base);
#else
    return is_signed? strtoll(str, endptr, base): strtoull(str, endptr, base);
#endif
}

w_base_t::int8_t	
w_base_t::strtoi8(
    const char *str,
    char **endptr,
    int 	base
)
{
    w_base_t::int8_t	i8;
    w_base_t::int8_t	u8 =
	    __strtou8(str, endptr, base, true);
    i8 = w_base_t::int8_t(u8);
    return i8;
}

w_base_t::uint8_t	
w_base_t::strtou8(
    const char *str,
    char **endptr,
    int 	base
)
{
    return __strtou8(str, endptr, base, false);
}

#ifdef _MSC_VER
#include <float.h>
#elif defined(SOLARIS2)
#include <ieeefp.h>
#else
#include <math.h>
#endif

bool
w_base_t::is_finite(const f8_t x)
{
    bool value = false;
#ifdef _MSC_VER
    value = _finite(x);
#else
    value = finite(x);
#endif
    return value;
}

bool
w_base_t::is_infinite(const f8_t x)
{
    bool value = false;
#ifdef _MSC_VER
    value = !_finite(x) && !_isnan(x);
#elif defined(SOLARIS2)
    value = !finite(x) && !isnand(x);
#else
    value = !finite(x) && !isnan(x);
#endif
    return value;
}

bool
w_base_t::is_nan(const f8_t x)
{
    bool value = false;
#ifdef _MSC_VER
    value = _isnan(x);
#elif defined(SOLARIS2)
    value = isnand(x);
#else
    value = isnan(x);
#endif
    return value;
}

bool
w_base_t::is_infinite_or_nan(const f8_t x)
{
    bool value = false;
#ifdef _MSC_VER
    value = !_finite(x);
#else
    value = !finite(x);
#endif
    return value;
}


void	w_base_t::abort()
{
	cout.flush();
	cerr.flush();
#ifdef _WIN32
	/* It would be great if there was a USER ERROR exception,
	   or something like that.  Lacking that, I'll steal 
	   illegal instruction and make it non-continuable.
	   If someone thinks of a better exception to throw, 
	   tell me about it, I'm all ears. Bolo */

	RaiseException(EXCEPTION_ILLEGAL_INSTRUCTION,
		       EXCEPTION_NONCONTINUABLE,
		       0, 0);
	/* Fall through to abort in case excpetion handling goes west */
#endif
	::abort();
}


/* Provide our own pure_virtual() so we can generate abort's if
   a virtual function is abused.  Name choice is somewhat
   hacky, since it is system and compiler dependent. */

#ifdef __GNUG__
#define	PURE_VIRTUAL	extern "C" void __pure_virtual()
#elif defined(_MSC_VER)
#undef _purecall
#define	PURE_VIRTUAL	_cdecl _purecall(void)
#else
#define	PURE_VIRTUAL	void pure_virtual()
#endif

PURE_VIRTUAL
{
	/* Just in case the iostreams generate another error ... */
	static	bool	called = false;
	if (!called)
		cerr << "** Pure virtual function called" << endl;
	called = true;

	w_base_t::abort();
	/*NOTREACHED*/

#ifdef _MSC_VER
	/* A hack  for visual c++ + inbred header files */
	return 0;
#endif
}
