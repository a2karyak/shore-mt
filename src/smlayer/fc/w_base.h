/*<std-header orig-src='shore' incl-file-exclusion='W_BASE_H'>

 $Id: w_base.h,v 1.60 1999/10/25 05:53:18 bolo Exp $

SHORE -- Scalable Heterogeneous Object REpository

Copyright (c) 1994-99 Computer Sciences Department, University of
                      Wisconsin -- Madison
All Rights Reserved.

Permission to use, copy, modify and distribute this software and its
documentation is hereby granted, provided that both the copyright
notice and this permission notice appear in all copies of the
software, derivative works or modified versions, and any portions
thereof, and that both notices appear in supporting documentation.

THE AUTHORS AND THE COMPUTER SCIENCES DEPARTMENT OF THE UNIVERSITY
OF WISCONSIN - MADISON ALLOW FREE USE OF THIS SOFTWARE IN ITS
"AS IS" CONDITION, AND THEY DISCLAIM ANY LIABILITY OF ANY KIND
FOR ANY DAMAGES WHATSOEVER RESULTING FROM THE USE OF THIS SOFTWARE.

This software was developed with support by the Advanced Research
Project Agency, ARPA order number 018 (formerly 8230), monitored by
the U.S. Army Research Laboratory under contract DAAB07-91-C-Q518.
Further funding for this work was provided by DARPA through
Rome Research Laboratory Contract No. F30602-97-2-0247.

*/

#ifndef W_BASE_H
#define W_BASE_H

#include "w_defines.h"

/*  -- do not edit anything above this line --   </std-header>*/

/*******************************************************/
/* get configuration definitions from config/shore.def */
/* (in w_defines.h above).                             */
#include <w_windows.h>
/*
 * WARNING: we turned on ON and OFF in shore.def
 * but we MUST turn them off as soon as configuration
 * stuff has been read, because ON and OFF are re-defined
 * elsewhere as enums
 */
#ifdef ON
#undef ON
#endif

#ifdef OFF
#undef OFF
#endif
/* end configuration definitions                       */
/*******************************************************/


#if !defined(W_OS2) && !defined(_WINDOWS)
#define W_UNIX
#endif

#ifdef __GNUG__
#pragma interface
#endif

#include <stddef.h>
#include <stdlib.h>
#include <w_stream.h>

#include <limits.h>

#if defined(W_UNIX) && !defined(_WIN32)
#include <unistd.h>
#endif

#if defined(_MSC_VER) && defined(small)
/* XXX namespace corruption in MSC header files */
#undef small
#endif

#ifndef W_WORKAROUND_H
#include "w_workaround.h"
#endif

#if defined(DEBUG) && !defined(W_DEBUG)
#define W_DEBUG
#endif

#define NORET		/**/
#define CAST(t,o)	((t)(o))
#define SIZEOF(t)	(sizeof(t))

#define	W_UNUSED(x)	/**/

#ifdef W_DEBUG
#define W_IFDEBUG(x)	x
#define W_IFNDEBUG(x)	/**/
#else
#define W_IFDEBUG(x)	/**/
#define W_IFNDEBUG(x)	x
#endif

#ifdef W_TRACE
#define	W_IFTRACE(x)	x
#define	W_IFNTRACE(x)	/**/
#else
#define	W_IFTRACE(x)	/**/
#define	W_IFNTRACE(x)	x
#endif

#if defined(W_DEBUG_SPACE)
void 	w_space(int line, const char *file);
#define W_SPACE w_space(__LINE__,__FILE__)
#else
#define W_SPACE
#endif


#define w_assert1(x)    do {						\
	if (!(x)) w_base_t::assert_failed(#x, __FILE__, __LINE__);	\
	W_SPACE;							\
} while(0)

#ifdef W_DEBUG
#define w_assert3(x)	w_assert1(x)
#else
#define w_assert3(x)	/**/
#endif

/*
 * Cast to treat an enum as in integer value.  This is used when
 * a operator<< doesn't exist for the enum.  The use of the macro
 * indicates that this enum would be printed if it had a printer,
 * rather than wanting the integer value of the enum
 */
#define	W_ENUM(x)	((int)x)

class w_rc_t;

/*--------------------------------------------------------------*
 *  w_base_t							*
 *--------------------------------------------------------------*/
class w_base_t {
public:
    /*
     *  shorthands
     */
    typedef unsigned char	u_char;
    typedef unsigned short	u_short;
    typedef unsigned long	u_long;
    // typedef w_rc_t		rc_t;

    /*
     *  basic types
     */
    typedef char		int1_t;
    typedef u_char		uint1_t;
    typedef short		int2_t;
    typedef u_short		uint2_t;
    typedef long		int4_t;
    typedef u_long		uint4_t;

#ifdef _MSC_VER
#    if defined(_INTEGRAL_MAX_BITS) && (_INTEGRAL_MAX_BITS >= 64)
    	typedef __int64 		int8_t;
    	typedef unsigned __int64 	uint8_t;
#    else
#	error Compiler does not support int8_t (__int64).
#    endif

#elif defined(__GNUG__)
    	typedef long long		int8_t;
    	typedef unsigned long long	uint8_t;
#else
#error int8_t Not supported for this compiler.
#endif

    /* 
     * For statistics that are always 64-bit numbers
     */
    typedef uint8_t 		large_stat_t; 

    /* 
     * For statistics that are 64-bit numbers 
     * only when #defined LARGEFILE_AWARE
     */
#ifdef LARGEFILE_AWARE
    typedef uint8_t     	base_stat_t;
#else
    typedef uint4_t     	base_stat_t;
#endif

    typedef float		f4_t;
    typedef double		f8_t;

    static const int1_t		int1_max, int1_min;
    static const int2_t		int2_max, int2_min;
    static const int4_t		int4_max, int4_min;
    static const int8_t		int8_max, int8_min;

    static const uint1_t	uint1_max, uint1_min;
    static const uint2_t	uint2_max, uint2_min;
    static const uint4_t	uint4_max, uint4_min;
    static const uint8_t	uint8_max, uint8_min;

    /*
     *  miscellaneous
     */
    enum { align_on = 0x8, align_mod = align_on - 1 };

	/*
	// NEH: turned into a macro for the purpose of folding
    // static uint4_t		align(uint4_t sz);
	*/
#ifndef align
#define alignonarg(a) (((uint4_t)a)-1)
#define alignon(p,a) (((uint4_t)((char *)p + alignonarg(a))) & ~alignonarg(a))

#define ALIGNON 0x4
#define ALIGNON1 (ALIGNON-1)
#define align(sz) ((uint4_t)((sz + ALIGNON1) & ~ALIGNON1))
#endif /* align */

    static bool		is_aligned(uint4_t sz);
    static bool		is_aligned(const void* s);

    static bool		is_big_endian();
    static bool		is_little_endian();

    /*  strtoi8 and strtou8 act like strto[u]ll with the following
     *  two exceptions: the only bases supported are 0, 8, 10, 16;
     *  ::errno is not set
     */
    static int8_t	strtoi8(const char *, char ** end=0 , int base=0);
    static uint8_t	strtou8(const char *, char ** end=0, int base=0);

    static istream&	_scan_uint8(istream& i, uint8_t &, 
				bool chew_white,
				bool is_signed,
				bool& rangerr);

    static bool		is_finite(const f8_t x);
    static bool		is_infinite(const f8_t x);
    static bool		is_nan(const f8_t x);
    static bool		is_infinite_or_nan(const f8_t x);

    /*
     *  standard streams
     */
    friend ostream&		operator<<(
	ostream&		    o,
	const w_base_t&		    obj);

    static void			assert_failed(
    	const char*		    desc,
	const char*		    file,
	uint4_t 		    line);

    static	void		abort();
};

/* XXX compilers+environment that need this operator defined */
#if defined(_MSC_VER)  
ostream &operator<<(ostream &o, const w_base_t::int8_t &t);
ostream &operator<<(ostream &o, const w_base_t::uint8_t &t);
istream &operator>>(istream &i, w_base_t::int8_t &t);
istream &operator>>(istream &i, w_base_t::uint8_t &t);
#endif

/*--------------------------------------------------------------*
 *  w_base_t::align()						*
 *--------------------------------------------------------------*/
/*
 * NEH: turned into a macro
inline w_base_t::uint4_t
w_base_t::align(uint4_t sz)
{
    return (sz + ((sz & align_mod) ? 
		  (align_on - (sz & align_mod)) : 0));
}
*/

/*--------------------------------------------------------------*
 *  w_base_t::is_aligned()					*
 *--------------------------------------------------------------*/
inline bool
w_base_t::is_aligned(uint4_t sz)
{
    return (align(sz) == sz);
}

inline bool
w_base_t::is_aligned(const void* s)
{
    return is_aligned(CAST(uint4_t, s));
}

/*--------------------------------------------------------------*
 *  w_base_t::is_big_endian()					*
 *--------------------------------------------------------------*/
inline bool w_base_t::is_big_endian()
{
	/* Provide fast results on known architectures, slower
	   but correct results on unknown architectures.
	 */
#if defined(Sparc)
	return true;
#elif defined(I386)
	return false;
#else
	int i = 1;
	return ((char*)&i)[3] == i;
#endif
}

/*--------------------------------------------------------------*
 *  w_base_t::is_little_endian()				*
 *--------------------------------------------------------------*/
inline bool
w_base_t::is_little_endian()
{
    return ! is_big_endian();
}

/*--------------------------------------------------------------*
 *  w_vbase_t							*
 *--------------------------------------------------------------*/
class w_vbase_t : public w_base_t {
public:
    NORET			w_vbase_t()	{};
    virtual NORET		~w_vbase_t()	{};
};

class w_msec_t {
public:
    enum special_t { 
	t_infinity = -1,
	t_forever = t_infinity,
	t_immediate = 0
    };

    NORET			w_msec_t();
    NORET			w_msec_t(w_base_t::uint4_t ms);
    NORET			w_msec_t(special_t s);
    
    bool			is_forever();
    w_base_t::uint4_t		value();
private:
    w_base_t::int4_t		ms;
};

inline w_base_t::uint4_t
w_msec_t::value()
{
    return CAST(w_base_t::uint4_t, ms);
}

inline bool
w_msec_t::is_forever()
{
    return ms == t_forever;
}

/*
 * These types are auto initialized filler space for alignment
 * in structures.  The auto init helps with purify.
 *
 * XXX Some of the users of these structures DEPEND on zero
 * fill (sm keys, etc). Eventually this will be seperated into
 * zero fill and plain fill (zeroed with PURIFY_ZERO).  Until then
 * these must always be initialized to 0.
 */
struct fill1 {
	w_base_t::uint1_t u1;
	fill1() : u1(0) {}
};

struct fill2 {
	w_base_t::uint2_t u2;
	fill2() : u2(0) {}
};

struct fill3 {
	w_base_t::uint1_t	u1[3];
	fill3() { u1[0] = u1[1] = u1[2] = 0; }
};

struct fill4 {
	w_base_t::uint4_t u4;
	fill4() : u4(0) {}
};


#include <w_autodel.h>
#include <w_factory.h>
#include <w_factory_fast.h>
#include <w_factory_thread.h>
#include <w_error.h>
#include <w_rc.h>

/*<std-footer incl-file-exclusion='W_BASE_H'>  -- do not edit anything below this line -- */

#endif          /*</std-footer>*/
