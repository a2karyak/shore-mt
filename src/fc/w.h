/* --------------------------------------------------------------- */
/* -- Copyright (c) 1994, 1995 Computer Sciences Department,    -- */
/* -- University of Wisconsin-Madison, subject to the terms     -- */
/* -- and conditions given in the file COPYRIGHT.  All Rights   -- */
/* -- Reserved.                                                 -- */
/* --------------------------------------------------------------- */

/*
 *  $Id: w.h,v 1.13 1995/04/24 19:31:45 zwilling Exp $
 */
#ifndef W_H
#define W_H

#include <w_base.h>
#include <w_minmax.h>
#include <w_list.h>
#include <w_hash.h>

#ifdef W_INCL_SHMEM
#include <w_shmem.h>
#endif

#ifdef W_INCL_CIRCULAR_QUEUE
#include <w_cirqueue.h>
#endif

#ifdef W_INCL_TIMER
#include <w_timer.h>
#endif

#ifdef W_INCL_BITMAP
#include <w_bitmap.h>
#endif

#endif /*W_H*/
