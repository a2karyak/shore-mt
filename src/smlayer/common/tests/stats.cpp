/*<std-header orig-src='shore'>

 $Id: stats.cpp,v 1.14 2003/06/19 18:43:51 bolo Exp $

SHORE -- Scalable Heterogeneous Object REpository

Copyright (c) 1994-99 Computer Sciences Department, University of
                      Wisconsin -- Madison
All Rights Reserved.

Permission to use, copy, modify and distribute this software and its
documentation is hereby granted, provided that both the copyright
notice and this permission notice appear in all copies of the
software, derivative works or modified versions, and any portions
thereof, and that both notices appear in supporting documentation.

THE AUTHORS AND THE COMPUTER SCIENCES DEPARTMENT OF THE UNIVERSITY
OF WISCONSIN - MADISON ALLOW FREE USE OF THIS SOFTWARE IN ITS
"AS IS" CONDITION, AND THEY DISCLAIM ANY LIABILITY OF ANY KIND
FOR ANY DAMAGES WHATSOEVER RESULTING FROM THE USE OF THIS SOFTWARE.

This software was developed with support by the Advanced Research
Project Agency, ARPA order number 018 (formerly 8230), monitored by
the U.S. Army Research Laboratory under contract DAAB07-91-C-Q518.
Further funding for this work was provided by DARPA through
Rome Research Laboratory Contract No. F30602-97-2-0247.

*/

#include "w_defines.h"

/*  -- do not edit anything above this line --   </std-header>*/

#include <stdio.h>
#include <stdlib.h>
#include <w_strstream.h>
#include <iostream.h>
#include <stddef.h>
#include <limits.h>
#include <w.h>
#include <w_base.h>
#include <option.h>
#include <w_debug.h>

#include "w_statistics.h"
#include "test_stat.h"

void statstest();

int main()
{

	DBG(<<"app_test: main");
	statstest();
	return 0;
}

#ifdef __GNUC__
typedef w_auto_delete_array_t<char> gcc_kludge_1;
typedef w_list_i<option_t> 			gcc_kludge_0;
#endif /* __GNUC__*/


void
statstest()
{
	/*********w_statistics_t ***********/

	class w_statistics_t ST; // generic stats class
	class test_stat TST; // my test class that uses stats
	w_rc_t	e;

	// server side: implementor of RPC request does this:
	TST.compute();

	ST << TST;
	ST << TST;
	ST << TST;

	// then ST gets sent across the wire and re-instantiated
	// on client side

	// Various ways the contents of ST can be printed:
	{
		cout << "ST.printf():" << endl;
		ST.printf();
		cout << endl;

		cout << "cout << ST:" << endl;
		cout << ST;
		cout << endl;

		/*  
		 * Format some output my own way.
		 * This is a way to do it by calling
		 * on the w_statistics_t structure to provide
		 * the string name.
		 */
		cout << "My own pretty formatting for module : " << ST.module(TEST_i) << endl;
		W_FORM2(cout,("\t%-30.30s %10.10d", 
				ST.string(TEST_i), ST.int_val(TEST_i)) );
			cout << endl;
		W_FORM2(cout,("\t%-30.30s %10.10u", 
				ST.string(TEST_j), ST.uint_val(TEST_j)) );
			cout << endl;
		W_FORM2(cout,("\t%-30.30s %10.6f", 
				ST.string(TEST_k), ST.float_val(TEST_k)) );
			cout << endl;
#ifdef notyet
		W_FORM2(cout,("\t%-30.30s %10.10d", 
				ST.string(TEST_l), ST.float_val(TEST_l)) );
			cout << endl;
#endif
		W_FORM2(cout,("\t%-30.30s %10.10u", 
				ST.string(TEST_v), ST.ulong_val(TEST_v)) );
			cout << endl;
		cout << endl;

		/*
		 * Error cases:
		 */
		cout << "Expect some unknowns-- these are error cases:" <<endl;
		cout <<  ST.typechar(TEST_i) << ST.typestring(TEST_STATMIN) << endl;
		cout 	// <<  ST.typechar(-1,1) 
			<< ST.typestring(-1) 
				<< ST.string(-1,-1)<<  ST.float_val(-1,-1)
		<< endl;
	}
	{
		w_statistics_t *firstp = ST.copy_brief();
		w_statistics_t &first = *firstp;

		if(!firstp) {
			cout << "Could not copy_brief ST." << endl;
		
		} else {
			const w_statistics_t &CUR = ST;

			cout << __LINE__ <<  " :******* CUR.copy_brief(true): " << endl;
			first.printf();
			cout << endl;

			cout << __LINE__ << endl;
			TST.inc();
			cout << __LINE__<< " :******* TST.inc(): " << endl;
			CUR.printf();
			cout << endl;
		}
		{
			w_statistics_t *lastp = ST.copy_all();
			w_statistics_t &last = *lastp;

			if(!lastp) {
				cout << "Could not copy_all ST." << endl;
			
			} else {
				cout << __LINE__ << endl;

				cout << __LINE__ << endl;
				last -= first;
				cout << __LINE__ 
					<< " :******* last -= first: :" << endl;
				cout << last << endl;
				cout << endl;

				cout << __LINE__ << endl;
				last += first;
				cout << __LINE__ 
					<< " :******* last += first: :" << endl;
				cout << last << endl;
				cout << endl;

				cout << __LINE__ << endl;
				last.zero();
				cout << __LINE__ 
					<< " :******* last.zero(): :" << endl;
				cout << last << endl;
				cout << endl;

				cout << "Expect error:" << endl;

				ST -= first;
			}
			delete lastp;
		}
		delete firstp;
	}
	/********** end w_statistics_t tests ********/
}

