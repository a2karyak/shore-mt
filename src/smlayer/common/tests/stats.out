ST.printf():

example statistics module : 
                 sample int:        300
        sample unsigned int:        819
               sample float:    1.23450
       sample unsigned long:         -1
  sample sum of i, j, and k: 4.29497e+09

cout << ST:

example statistics module : 
                 sample int:        300
        sample unsigned int:        819
               sample float:    1.23450
       sample unsigned long:         -1
  sample sum of i, j, and k: 4.29497e+09

My own pretty formatting for module : example statistics module 
	sample int                     0000000300
	sample unsigned int            0000000819
	sample float                     1.234500
	sample unsigned long           -1074790400

Expect some unknowns-- these are error cases:
iiufvf
unknown-1
134 :******* CUR.copy_brief(true): 

example statistics module : 
                 sample int:        300
        sample unsigned int:        819
               sample float:    1.23450
       sample unsigned long:         -1
  sample sum of i, j, and k: 4.29497e+09

138
140 :******* TST.inc(): 

example statistics module : 
                 sample int:        301
        sample unsigned int:        820
               sample float:    2.23450
  sample sum of i, j, and k:    1123.23

152
154
156 :******* last -= first: :

example statistics module : 
                 sample int:          1
        sample unsigned int:          1
               sample float:    1.00000
       sample unsigned long:          1
  sample sum of i, j, and k: -4.29497e+09


161
163 :******* last += first: :

example statistics module : 
                 sample int:        301
        sample unsigned int:        820
               sample float:    2.23450
  sample sum of i, j, and k:    1024.00


168
170 :******* last.zero(): :

example statistics module : 


Expect error:
